Analyze command

edwinkortman opened the pull request:
Fixes #56.

So, ehm. It's a bit of a mess because I merged in https://github.com/corgibytes/freshli-cli/pull/224 and then had to make some changes to it because I forgot something there. My apologies :disguised_face:  

>  Adjust number of workers (this could be considered a nice to have if it turns out to be difficult)

So I've tried to do this but I don't think this is possible without recreating the entire Hangfire server. 
What I had hoped would work is adjust the background server job options but I couldn't confirm an increase in workers.

```csharp
public AnalyzeRunner(
        IServiceProvider serviceProvider,
        Runner runner,
        IApplicationActivityEngine activityEngine,
        ICacheManager cacheManager,
        BackgroundJobServerOptions backgroundJobServerOptions,
        IHostedService backgroundJobServer
        ) : base(serviceProvider, runner)
    {
        _activityEngine = activityEngine;
        _cacheManager = cacheManager;
        _backgroundJobServerOptions = backgroundJobServerOptions;
        _backgroundJobServer = backgroundJobServer;
    }
...

_backgroundJobServerOptions.WorkerCount = 25;
_backgroundJobServer.StopAsync(new CancellationToken());
_backgroundJobServer.StartAsync(new CancellationToken());
```

mscottford commented:
Let's move controlling the worker count via a parameter to a separate issue, since there were challenges getting it to work.

mscottford commented:
Are #205 and #208 also fixed by this pull request? If so, we should correctly note that in the description.

mscottford commented on Corgibytes.Freshli.Cli.Test/CommandOptions/AnalyzeCommandOptionsTest.cs:
```diff
@@ -0,0 +1,104 @@
+
+        var result = parser.Parse(args);
+
+        var gitPath = result.GetOptionValueByName<string>("git-path");
+        gitPath?.Should().NotBeEmpty().And.Be(expectedGitPath);
```
If `gitPath` is null then there's no assertion made against it's value. That doesn't seem correct to be. Seems like it doesn't properly test it's default value.

edwinkortman replied:
That's definitely not correct. I didn't realize that. Thanks! 
The only value that is nullable is branch. https://github.com/corgibytes/freshli-cli/pull/212/commits/cb9a824809fa09908059b1779e63e790c6262be0 this commit reflects that.

mscottford commented on Corgibytes.Freshli.Cli.Test/CommandOptions/AnalyzeCommandOptionsTest.cs:
```diff
@@ -0,0 +1,104 @@
+        var gitPath = result.GetOptionValueByName<string>("git-path");
+        gitPath?.Should().NotBeEmpty().And.Be(expectedGitPath);
+
+        var branch = result.GetOptionValueByName<string>("branch");
+        branch?.Should().NotBeEmpty().And.Be(expectedBranch);
```
Similar to the comment about `gitPath`. If the value is `null` then there's no assertion made against it's value.

edwinkortman replied:
Addressed in https://github.com/corgibytes/freshli-cli/pull/212/commits/cb9a824809fa09908059b1779e63e790c6262be0

mscottford commented on Corgibytes.Freshli.Cli.Test/CommandOptions/AnalyzeCommandOptionsTest.cs:
```diff
@@ -0,0 +1,104 @@
+        var commitHistory = result.GetOptionValueByName<bool>("commit-history");
+        commitHistory.Should().Be(expectedCommitHistory);
+
+        var historyInterval = result.GetOptionValueByName<string>("history-interval");
+        historyInterval?.Should().NotBeEmpty().And.Be(expectedHistoryInterval);
```
Same as `gitPath` and `branch`. A `null` value for `historyInterval` means that it doesn't get asserted.

edwinkortman replied:
Addressed in https://github.com/corgibytes/freshli-cli/pull/212/commits/cb9a824809fa09908059b1779e63e790c6262be0

mscottford commented on Corgibytes.Freshli.Cli.Test/CommandOptions/AnalyzeCommandOptionsTest.cs:
```diff
@@ -0,0 +1,104 @@
+    [Fact]
+    public void Send_Args_ReturnsAnalyzeArguments()
+    {
+        var cmBuilder = Program.CreateCommandLineBuilder();
```
I'm not a huge fan of the `cm` prefix on `cmBuilder`. I think just `builder` would be more than fine in this context, or the call could be inlined and removed altogether.

mscottford commented on Corgibytes.Freshli.Cli.Test/Functionality/CacheManagerTest.cs:
```diff
@@ -32,7 +32,7 @@ public void SavePersistsACachedAnalysisAndGeneratesAnId()
         var cache = cacheManager.GetCacheDb(_tempCacheDir);
 
-        var expectedAnalysis = new CachedAnalysis("https://git.example.com", "main", "1m");
+        var expectedAnalysis = new CachedAnalysis("https://git.example.com", "main", "1m", true);
```
I'm not a big fan of boolean values being passed as parameters, because it's always hard to tell what they mean. Can we either change all of these calls to look like:

```suggestion
        var expectedAnalysis = new CachedAnalysis("https://git.example.com", "main", "1m", useCommitHistory: true);
```

Or switch to using an enumeration. Something like `CommitHistory.Full` vs `CommitHistory.Minimal`, where `CommitHistory.Full` is analogous to a value of `true` and `CommitHistory.Minimal` is analogous to a value of `false`.

mscottford commented:
From reading through the changes, it looks like this pull request also addresses #213.

mscottford commented on Corgibytes.Freshli.Cli/CommandRunners/AnalyzeRunner.cs:
```diff
@@ -0,0 +1,65 @@
+                libYearSummed = computedEvent.LibYearPackages.Sum(libYear => libYear.LibYear);
+            }
+
+            context.Console.Out.WriteLine($"Libyear at {computedEvent.AnalysisLocation?.CommitId} is {libYearSummed}");
```
I like that this information is being written out. It'll help show progress as the command is running along. I think the output would be more meaningful if it included the path to the manifest that's being analyzed. Many git repositories are going to have more than one dependency manifest. In those cases, you'd get this output twice for a given CommitId. Also the date maters, too... so that should be included as well. So I think the output would need to be something like

"LibYear for {manifest} at {CommitId} on {date} is {libYearSummed}". What do you think? We can also create an issue for this and have it wait until later. I wouldn't think that this is needed before we demo.

We could also write this line to the log as `Information` if it's going to make the console output too noisy. If we write a detailed message to the log, then I think we could just write out a `.` to the console to show that progress is being made.

edwinkortman replied:
Haha, yeah I had the same conclusion but I didn't want to add it in this PR. Same with the acceptance tests, I wanted to wait until everything is connected before finalizing them as the tests depend on this output. 

I like the idea of logging it to a log instead of the console output. Maybe add a `--verbose` option to see these sort of logs?

edwinkortman replied:
Continuing in issue: https://github.com/corgibytes/freshli-cli/issues/240

edwinkortman replied:
Also related: https://github.com/corgibytes/freshli-cli/issues/241

mscottford commented on Corgibytes.Freshli.Cli/CommandRunners/ComputeLibYearCommandRunner.cs:
```diff
@@ -40,8 +41,11 @@ public override int Run(ComputeLibYearCommandOptions options, InvocationContext
+        _activityEngine.Dispatch(new ComputeLibYearActivity(
+            _calculateLibYearFromCycloneDxFile,
+            options.FilePath.ToString(),
+            new AnalysisLocation("this-runner-will-be-removed", "this-runner-will-be-removed")
```
Should the runner be removed now?

edwinkortman replied:
Yup! I totally forgot about that.

mscottford commented on Corgibytes.Freshli.Cli/CommandRunners/LoadServiceCommandRunner.cs:
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.CommandLine.Invocation;
+using Corgibytes.Freshli.Cli.CommandOptions;
+using Corgibytes.Freshli.Cli.Commands;
+using Corgibytes.Freshli.Cli.Functionality;
+using Corgibytes.Freshli.Cli.Functionality.Engine;
+using Corgibytes.Freshli.Lib;
+
+namespace Corgibytes.Freshli.Cli.CommandRunners;
+
+public class LoadServiceCommandRunner : CommandRunner<LoadServiceCommand, EmptyCommandOptions>
+{
+    public LoadServiceCommandRunner(IServiceProvider serviceProvider, Runner runner,
+        IApplicationActivityEngine activityEngine)
+        : base(serviceProvider, runner) =>
+        ActivityEngine = activityEngine;
+
+    private IApplicationActivityEngine ActivityEngine { get; }
+
+    public override int Run(EmptyCommandOptions options, InvocationContext context)
+    {
+        ActivityEngine.Dispatch(new LoadServiceProviderActivity());
+        ActivityEngine.Wait();
+        return 0;
+    }
+}
```
I really like using this to test that services are getting loaded correctly. Nicely done!

edwinkortman replied:
Thank you! :D

mscottford commented on Corgibytes.Freshli.Cli/CommandRunners/AnalyzeRunner.cs:
```diff
@@ -0,0 +1,65 @@
+
+        _activityEngine.Wait();
+
+        return 0;
```
There is no error checking going on in this method. Should there be? I would also expect us to return a non-zero result if the analysis did not correctly run to completion.

edwinkortman replied:
There should be, I guess we have to build something that can listen to any error event (instead of having to listen to each error event individually). I'll add an issue for it.

mscottford commented on Corgibytes.Freshli.Cli/Functionality/Git/CloneGitRepositoryActivity.cs:
```diff
@@ -13,45 +13,28 @@ public class CloneGitRepositoryActivity : IApplicationActivity
-            var gitRepository = _gitSourceRepository.CloneOrPull(_repoUrl, _branch, _cacheDir, _gitPath);
+            // Null warning suppressed, if we can't fetch it from the service container we got way bigger problems.
+            var gitRepository =
+                eventClient.ServiceProvider.GetService<ICachedGitSourceRepository>()!.CloneOrPull(_repoUrl, _branch,
```
You can use `GetRequiredService` to get rid of the `!`.

mscottford commented on features/analyze.feature:
```diff
@@ -0,0 +1,53 @@
+    With no options specified, performs analysis locally, and then sends the results to the Freshli web app so that the results can be viewed at at URL that will be provided in the command output.
+
+
+    NOTE: These scenarios are not complete nor properly tested because at the time of writing not everything is done. So the entire chain can't be tested yet.
```
What hasn't been done yet? We'll need to create issues for them, and also create an issue to ensure that these scenarios are fleshed out.

mscottford replied:
Also... why wouldn't it make sense to complete everything as part of this pull request?

edwinkortman replied:
I'll make sure to create the issue for it. 
Why it doesn't make sense to complete everything is that it depends on some other issues that aren't done yet:
https://github.com/corgibytes/freshli-cli/issues/205
https://github.com/corgibytes/freshli-cli/issues/208

I'm going to check the entire flow and see if we are missing any connections and make issues for it. After that's done we can update these features.

edwinkortman replied:
I've added issues and linked them to the epic here: 
https://github.com/corgibytes/freshli-cli/issues/135#issuecomment-1241804496
There's also an updated image of the happy path flow to show what we are missing because I lost the overview entirely and misguided Dona twice now.
https://github.com/corgibytes/freshli-cli/issues/135#issuecomment-1241805801

mscottford reviewed (dismissed):
Lots of work here! I have some comments and questions. Some minor. Some will likely lead to the creation of other issues.

edwinkortman commented:
> Let's move controlling the worker count via a parameter to a separate issue, since there were challenges getting it to work.

Added #232 to iteration 9 but as MVP nice-to-have. 


> Are #205 and #208 also fixed by this pull request? If so, we should correctly note that in the description.
> From reading through the changes, it looks like this pull request also addresses #213. If that's true, please also add a "Fixes..." line for it to the description of this pull request.

Nope, neither of those issues are fixed in this PR. I have made some changes to `CloneGitRepositoryActivity` but the work for those issues still need to be done.

mscottford commented:
> > Let's move controlling the worker count via a parameter to a separate issue, since there were challenges getting it to work.
> 
> Added #232 to iteration 9 but as MVP nice-to-have.

Yay!
 
> > Are #205 and #208 also fixed by this pull request? If so, we should correctly note that in the description.
> > From reading through the changes, it looks like this pull request also addresses #213. If that's true, please also add a "Fixes..." line for it to the description of this pull request.
> 
> Nope, neither of those issues are fixed in this PR. I have made some changes to `CloneGitRepositoryActivity` but the work for those issues still need to be done.

Assuming that this pull request gets merged in first, we'll need to strengthen the description of those issues so that acceptance tests get properly added to the `analyze` command along the way.

edwinkortman commented on Corgibytes.Freshli.Cli/CommandRunners/Git/GitCloneCommandRunner.cs:
```diff
@@ -15,22 +15,19 @@ public class GitCloneCommandRunner : CommandRunner<GitCloneCommand, GitCloneComm
     {
-        _activityEngine.Dispatch(new CloneGitRepositoryActivity(_gitSourceRepository,
-            options.RepoUrl, options.Branch, options.CacheDir, options.GitPath));
+        // RepositoryUrl, Branch, CacheDir, GitPath, AnalysisId
```
Whoops, that's what I sometimes use to get the parameters in the correct order so I don't have to check the method signature each time.

edwinkortman commented on Corgibytes.Freshli.Cli/Functionality/FreshliWeb/ResultsApi.cs:
```diff
@@ -0,0 +1,10 @@
+public class ResultsApi : IResultsApi
+{
+    public string GetResultsUrl(Guid analysisId) =>
+        // Not the final implementation
+        "https://freshli.app/" + analysisId;
```
Haha I know right.

edwinkortman commented:
> Assuming that this pull request gets merged in first, we'll need to strengthen the description of those issues so that acceptance tests get properly added to the analyze command along the way.

Yes, I agree but the feature test is depending on the entire chain to work before the output can be properly tested. 
Added an issue for this: https://github.com/corgibytes/freshli-cli/issues/240
